### DISCUSSION TITLE
Gyrotropic Media parameters

### DISCUSSION THREAD

**Discussion by cciso** [NanoComp/meep#2637] [General]
Good morning,

I was looking at the Gyrotropic media tutorial (https://meep.readthedocs.io/en/latest/Python_Tutorials/Gyrotropic_Media/ ) 
and I would like to know if the parameters chosen for the gyrotropic Lorentzian medium (epsn = 1.5    # background permittivity
f0   = 1.0    # natural frequency, gamma = 1e-6   # damping rate, sn   = 0.1    # sigma parameter, b0   = 0.15   ) correspond to a real medium or if they have been chosen randomly? 

There is no reference to a specific material and I am trying to understand how to put real life materials in the simulation which are often given in terms of Verdet constants, external magnetic field in Tesla... I find it difficult to make the connection with what Meeps does.

Thank you very much.

**Comment by cciso** [comment]
After much consideration, I think being able to input all components of the epsilon tensor would be more useful, however, as far as I can tell Meep only allows us to define off-diagonal elements but not individual elements, is there a reason for this? 

I found an old thread (https://meep-discuss.ab-initio.mit.narkive.com/tF1Ku4ca/anisotropy-in-meep#post2) which gave hints on how to do this with opened questions which I am interested in knowing the answers to: 

"The easy part was on the libctl side; in the specifications file I
modified dielectric materials to have "epsilon-tensor" and "scaled-
epsilon-tensor" properties. If you provide no tensor when making your
dielectric, the material is assumed to be isotropic and the scaled
tensor will have the same dielectic constant along its diagonal.

If you want an anisotropic material, you can set epsilon to 1 and set
the epsilon-tensor to whatever you choose:

(define-class dielectric material-type
(define-property epsilon no-default 'number)
(define-property epsilon-tensor (matrix3x3 (vector3 1 0 0) (vector3 0 1 0) (vector3 0 0 1)) 'matrix3x3)
(define-derived-property scaled-epsilon-tensor 'matrix3x3
(lambda (object)
(matrix3x3* (object-property-value object 'epsilon)
(object-property-value object 'epsilon-tensor))))
(define-property polarizations '() (make-list-type 'polarizability))
(define-property chi2 0 'number)
(define-property chi3 0 'number)
)

Resulting in this behavior:

meep> (define a (make dielectric (epsilon 3)))
meep> (define b (make dielectric (epsilon 1) (epsilon-tensor (matrix3x3 (vector3 4.8 0 0) (vector3 0 4.8 0) (vector3 0 0 5.2)))))
meep> a
((dielectric material-type) (chi3 . 0) (chi2 . 0) (polarizations) (epsilon-tensor . #(#(1 0 0) #(0 1 0) #(0 0 1))) (epsilon . 3) (scaled-epsilon-tensor . #(#(3 0 0) #(0 3 0) #(0 0 3))))
meep> b
((dielectric material-type) (chi3 . 0) (chi2 . 0) (polarizations) (epsilon-tensor . #(#(4.8 0 0) #(0 4.8 0) #(0 0 5.2))) (epsilon . 1) (scaled-epsilon-tensor . #(#(4.8 0 0) #(0 4.8 0) #(0 0 5.2))))


On the C side, things get trickier, and I'm a bit lost. There are
several things I don't understand in the code:

1) structure_chunk::set_epsilon (specified in
anisotropic_averaging.cpp) is passed a material_function. That class
has one-liner methods specified in meep.hpp, (eps(), sigma(), chi3(),
and chi2()).

public:
simple_material_function(double (*func)(const vec &)) { f = func; }

virtual ~simple_material_function() {}

virtual double eps(const vec &r) { return f(r); }
virtual double sigma(const vec &r) { return f(r); }
virtual double chi3(const vec &r) { return f(r); }
virtual double chi2(const vec &r) { return f(r); }
};


I think when using the c++ interface, you define these functions yourself, like this:

double eps(const vec &p) {
vec r = p - center;
if (abs(r) < rad) return er;
else return 1.0;
}

But how do these functions get informed or populated when running meep
via libctl? More specifically, how can material_functions access the
scaled-epsilon-tensor (matrix3x3) above?

2) Should a new subclass of material_function be defined? In
simple_material_function, eps, sigma, chi3, chi2 are all doubles, so
you can use that template-esque code above. Since the epsilon-tensor
is not a double, I should probably return a matrix3x3 struct (those
are typedef'ed in ctl.h, which I think meep.hpp is unaware of... ).

3) If I manage to set the inveps of structure_chunk, any subsequent
calls to set_epsilon() will clobber those values. This means the
interface should set inveps only after set_epsilon has been called?

4) Is it possible to set inveps in a way which takes advantage of the
existing anisotropic averaging code (which also sets inveps)?"

**Comment by stevengj** [comment]
> After much consideration, I think being able to input all components of the epsilon tensor would be more useful, however, as far as I can tell Meep only allows us to define off-diagonal elements but not individual elements, is there a reason for this?

Physically, you can't just put in a frequency-independent non-symmetric ε tensor into a time-domain simulation, because such a material would not be passive (it would lead to diverging fields).   It's for the same reason that you [can't just enter a negative ε](https://meep.readthedocs.io/en/latest/FAQ/#why-does-my-simulation-diverge-if-the-permittivity-is-less-than-0).

Such materials really *have* to be frequency dependent, which means that they need to correspond in time domain to some sort of model, like the ones implemented in Meep.

**Reply by cciso** [reply]
Thank you very much Stevengj, that makes sense!
